server/src/weaverbird/backends/utils/formula_builder.py

```diff
@@ -0,0 +1,566 @@
+"""
+
+A formula builder module for all our backends
+
+"""
+import ast
+import re
+from functools import lru_cache
+from typing import Any
+
+import numpy as np
+from pandas import DataFrame
+
+from weaverbird.pipeline.steps import FormulaStep, formula
+
+
+class InvalidFormula(Exception):
+    """
+    Raised when a formula is not supported
+    """
+
+
+class ColumnBuilder:
+    """
+    A column delemiter module
+
+    As default, the suffix is the same as the prefix
+    so that if prefix is ", we could end with "col"
+
+    Also called as QUOTE_CHAR, these characters (prefix/suffix) will delimit
+    our columns.
+
+    Examples:
+    In [2]: ColumnBuilder('"', 'test')
+    Out[2]: '"test"'
+
+    In [3]: ColumnBuilder('$', 'test')
+    Out[3]: '$test'
+
+    In [4]: ColumnBuilder('${', 'test')
+    Out[4]: '${test}'
+
+    In [5]: ColumnBuilder('[', 'test')
+    Out[5]: '[test]'
+    """
+
+    ENCLOSING = {'`': '`', '"': '"', '\'': '\'', '${': '}', '$(': ')', '[': ']', '{': '}', '(': ')'}
+
+    def __init__(self, prefix: str, column_name: str, suffix: str | None = None) -> None:
+        self.prefix = prefix
+        self.column_name = column_name
+        self.suffix = self._build_suffix(self.prefix, suffix)
+
+    def _build_suffix(self, prefix: str, suffix: str | None) -> str:
+        return self.ENCLOSING.get(prefix, '') if not suffix else suffix
+
+    def __str__(self) -> str:
+        """For a given column_name render with enclosures"""
+        return f'{self.prefix}{self.column_name}{self.suffix}'
+
+    def __repr__(self) -> str:
+        """For a given column_name render with enclosures"""
+        return f'{self.prefix}{self.column_name}{self.suffix}'
+
+    def __add__(self, other) -> str:
+        return self.__str__() + other
+
+    def __radd__(self, other) -> str:
+        return self.__str__() + other
+
+
+class FormulaBuilder:
```

Comment:
I'm not sure that the class helps much here, as there is no instance data. Maybe just exporting functions directly would be enough?